transit_odp/organisation/csv/service_codes.py

```diff
@@ -1,59 +1,251 @@
-        writer = csv.writer(string_, quoting=csv.QUOTE_ALL)
-        writer.writerow(
-            ("Service Status", "Licence Number", "Service Code", "Line", "Dataset ID")
+def get_all_otc_map(organisation_id: int) -> Dict[str, OTCService]:
+    """
+    Get a dictionary which includes all OTC Services for an organisation.
+    """
+    return {
+        service.registration_number.replace("/", ":"): service
+        for service in OTCService.objects.get_all_otc_data_for_organisation(
+            organisation_id
         )
+    }
+
+
+def get_seasonal_service_map(organisation_id: int) -> Dict[str, SeasonalService]:
+    """
+    Get a dictionary which includes all the Seasonal Services
+    for an organisation.
+    """
+    return {
+        service.registration_number.replace("/", ":"): service
+        for service in SeasonalService.objects.filter(
+            licence__organisation_id=organisation_id
+        )
+        .add_registration_number()
+        .add_seasonal_status()
+    }
+
 
-        for otc_service in self.get_queryset():
-            row = self.get_row(otc_service)
-            if row[0] == self.MISSING_DATA:
-                writer.writerow(row)
-
-        for otc_service in self.get_queryset():
-            row = self.get_row(otc_service)
-            if row[0] == self.REGISTERED:
-                writer.writerow(row)
-
-        string_.seek(0)
-        return string_
-
-    def get_row(self, otc_service: OTCService) -> Tuple:
-        datasets: Set = {
-            str(dataset_id)
-            for service_code, dataset_id in self.org_service_code_to_dataset
-            if service_code == otc_service.service_code
-        }
-        return (
-            self.REGISTERED if len(datasets) > 0 else self.MISSING_DATA,
-            otc_service.licence.number,
-            otc_service.service_code,
-            otc_service.service_number,
-            ";".join(datasets),
+class ServiceCodesCSV(CSVBuilder):
+    columns = [
+        CSVColumn(
+            header="Requires Attention",
+            accessor=lambda otc_service: otc_service.get("require_attention"),
+        ),
+        CSVColumn(
+            header="Published Status",
+            accessor=lambda otc_service: "Published"
+            if otc_service.get("dataset_id")
+            else "Unpublished",
+        ),
+        CSVColumn(
+            header="OTC status",
+            accessor=lambda otc_service: "Registered"
+            if otc_service.get("licence_number")
+            else "Unregistered",
+        ),
+        CSVColumn(
+            header="Scope Status",
+            accessor=lambda otc_service: "Out of Scope"
+            if otc_service.get("scope_status")
+            else "In Scope",
+        ),
+        CSVColumn(
+            header="Seasonal Status",
+            accessor=lambda otc_service: add_seasonal_service_status(otc_service)
+            if otc_service.get("seasonal_status") is not None
+            else None,
```

Comment:
The `add_seasonal_service_status` function (which I've requested you rename above!) actually gives you what you want to output here. You shouldn't ever be outputting `None`.

Reply:
`seasonal_status` return a `bool` can be `True` or `False` if there is an `SeasonalService` for that `service_code`. This export is about `all` from `OTCService` and `TXCFileAttributes`. If there is no `seasonal_service` this row should be `None`. I think I have to leave as it is